SBTi/data/sbti.py

```diff
@@ -22,30 +24,51 @@ def __init__(
             output.write(resp.content)
             print(f'Status code from fetching the CTA file: {resp.status_code}, 200 = OK')
         # Read CTA file into pandas dataframe
+        # Suppress warning about openpyxl - check if this is still needed in the released version.
+        warnings.filterwarnings('ignore', category=UserWarning, module='openpyxl')
         self.targets = pd.read_excel(self.c.FILE_TARGETS)
+        
+    
+    def filter_cta_file(self, targets):
+        """
+        Filter the CTA file to create a datafram that has on row per company with the columns "Action" and "Target"
+        If Action = Target then only keep the rows where Target = Near-term.
+        Validate ISIN and LEI.
+        """
+        # Regex patterns for ISIN and LEI. May be used at a later stage.
+        isin_regex = r'^[A-Z]{2}[A-Z0-9]{9}[0-9]$'
+        lei_regex = r'^[0-9A-Z]{18}[0-9]{2}$' # A very simple regex for LEI. Not perfect but should be good enough.
 
-    # def get_sbti_targets(
-    #     self, companies: List[IDataProviderCompany], isin_map: dict
-    # ) -> List[IDataProviderCompany]:
-    #     """
-    #     Check for each company if they have an SBTi validated target.
-
-    #     :param companies: A list of IDataProviderCompany instances
-    #     :param isin_map: A map from company id to ISIN
-    #     :return: A list of IDataProviderCompany instances, supplemented with the SBTi information
-    #     """
-    #     for company in companies:
-    #         targets = self.targets[
-    #             self.targets[self.c.COL_COMPANY_ISIN]
-    #             == isin_map.get(company.company_id)
-    #         ]
-    #         if len(targets) > 0:
-    #             company.sbti_validated = (
-    #                 self.c.VALUE_TARGET_SET in targets[self.c.COL_TARGET_STATUS].values
-    #             )
+        # Create a new dataframe with only the columns "Action" and "Target"
+        targets = targets[[self.c.COL_COMPANY_NAME, self.c.COL_COMPANY_ISIN, self.c.COL_COMPANY_LEI, self.c.COL_ACTION, self.c.COL_TARGET]]
+        #
+        # Keep rows where Action = Target and Target = Near-term 
+        df = targets[(targets[self.c.COL_ACTION] == self.c.VALUE_ACTION_TARGET) & (targets[self.c.COL_TARGET] == self.c.VALUE_TARGET_SET)]
+        # Drop duplicates in the dataframe by waterfall. Do company name last due to risk of misspelled names
+        df = pd.concat([df[~df[self.c.COL_COMPANY_LEI].isnull()].drop_duplicates(subset=self.c.COL_COMPANY_LEI, keep='first'), df[df[self.c.COL_COMPANY_LEI].isnull()]])
+        df = pd.concat([df[~df[self.c.COL_COMPANY_ISIN].isnull()].drop_duplicates(subset=self.c.COL_COMPANY_ISIN, keep='first'), df[df[self.c.COL_COMPANY_ISIN].isnull()]])
+        df.drop_duplicates(subset=self.c.COL_COMPANY_NAME, inplace=True)
 
-    #     return companies
+        """
+        Code to validate the ISIN and LEI. This is not needed for the regular users.
```

Comment:
You're right, best remove it. I added it as a check of the ISIN and LEI, but since there are so many faults in the CTA file, it is probably better to run those validations separately.